### PR TITLE
1. Fix some address crawler timing issues

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

Zebra's address crawler is fragile - small changes can cause significant regressions.

This PR deals with a few different sources of that instability.

## Solution

- Fix a race between the first heartbeat and getaddr requests
- Temporarily reduce the getaddr fanout to 1, until we implement #2214 
- Make sure finished requests are always sent as soon as they are ready
- Simplify the `Connection` awaiting request state machine
- Improve connection metrics and logging

## Review

This PR is ready for review.

Let's assign a reviewer when people are back from leave.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs

I've tested this PR locally, and the crawler and address book metrics look a lot better. The number of peer addresses rises rapidly, and gets full after a few hours. And Zebra doesn't hang any more, even after a few days.

It's hard to test this PR, because the timing issues are subtle, and don't show up in some network environments. We should have better tests after we implement #1592.

### CODE REVIEW

**Comment by teor2345** [comment]
Bumping this up to high priority, because this fixes some occasional CI failures, and it might conflict with other PRs.

**Comment by teor2345** [comment]
I've tested this PR locally, and the crawler and address book metrics look a lot better. The number of peer addresses rises rapidly, and gets full after a few hours. And Zebra doesn't hang any more, even after a few days.

File: zebra-network/src/peer/connection.rs [hunk 1]
```diff
@@ -593,32 +600,19 @@ where
 
                             self.update_state_metrics(None);
 
-                            // # Correctness
+                            // Check whether the handler is finished processing messages,
+                            // and update the state.
+                            // (Some messages can indicate that a response has finished,
+                            // even if the message wasn't consumed as a response or a request.)
```
Comment (dconnolly) [inline zebra-network/src/peer/connection.rs:606]: 👍

File: zebra-network/src/peer/connection.rs [hunk 2]
```diff
@@ -631,27 +625,36 @@ where
...
+                                None
+                            };
+
+                            if let Some(unused_msg) = unused_msg {
+                                debug!(
+                                    %unused_msg,
+                                    %self.state,
+                                    "ignoring peer message: not a response or a request",
+                                );
                             }
```
Comment (dconnolly) [inline zebra-network/src/peer/connection.rs:658]: 👍

File: zebra-network/src/peer/handshake.rs [hunk 3]
```diff
@@ -978,8 +983,16 @@ async fn send_periodic_heartbeats(
 ) {
     use futures::future::Either;
 
-    let mut interval_stream =
-        IntervalStream::new(tokio::time::interval(constants::HEARTBEAT_INTERVAL));
+    // Don't send the first heartbeat immediately - we've just completed the handshake!
```
Comment (dconnolly) [inline zebra-network/src/peer/handshake.rs:986]: 👍

File: zebra-network/src/peer/handshake.rs [hunk 4]
```diff
@@ -978,8 +983,16 @@ async fn send_periodic_heartbeats(
...
+    let mut interval = tokio::time::interval_at(
+        Instant::now() + constants::HEARTBEAT_INTERVAL,
+        constants::HEARTBEAT_INTERVAL,
+    );
+    // If the heartbeat is delayed, also delay all future heartbeats.
+    // (Shorter heartbeat intervals just add load, without any benefit.)
+    interval.set_missed_tick_behavior(tokio::time::MissedTickBehavior::Delay);
```
Comment (dconnolly) [inline zebra-network/src/peer/handshake.rs:993]: 👍

File: zebra-network/src/protocol/external/message.rs [hunk 5]
```diff
@@ -401,25 +403,39 @@ impl fmt::Display for Message {
...
-            Message::Block(_) => "block".to_string(),
+            Message::Block(block) => format!(
+                "block {{ height: {}, hash: {} }}",
+                block
+                    .coinbase_height()
+                    .as_ref()
+                    .map(|h| h.0.to_string())
+                    .unwrap_or_else(|| "None".into()),
+                block.hash(),
+            ),
```
Comment (dconnolly) [inline zebra-network/src/protocol/external/message.rs:433]: 👍

File: zebra-network/src/protocol/internal/response.rs [hunk 6]
```diff
@@ -79,7 +79,21 @@ impl fmt::Display for Response {
...
+                format!(
+                    "Block {{ height: {}, hash: {} }}",
+                    block
+                        .coinbase_height()
+                        .as_ref()
+                        .map(|h| h.0.to_string())
+                        .unwrap_or_else(|| "None".into()),
+                    block.hash(),
+                )
+            }
```
Comment (dconnolly) [inline zebra-network/src/protocol/internal/response.rs:94]: 👍

**Review by dconnolly — APPROVED** [review]
lgtm!

**Pull request merged**